File: 0908-middle-of-the-linked-list/0908-middle-of-the-linked-list.cpp
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        int count = 0;
        ListNode* temp = head;

        // Step 1: Count total nodes
        while (temp != nullptr) {
            count++;
            temp = temp->next;
        }

        // Step 2: Find middle index
        int mid = count / 2;  // If even, returns the second middle as per LeetCode definition

        // Step 3: Traverse again to the middle
        temp = head;
        for (int i = 0; i < mid; i++) {
            temp = temp->next;
        }

        return temp;
    }
};

File: 0908-middle-of-the-linked-list/0908-middle-of-the-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode middleNode(ListNode head) {
        if( head == null || head.next == null) return head; 
        int l = 0; 
        ListNode temp = head; 
        while( temp!= null){
            l+=1; 
            temp = temp.next; 
        }
        int middle = l/2; 
        temp = head; 

        for( int i = 0; i< middle ; i++){
            temp = temp.next; 
        }
        return temp;
        
    }
}